### ISSUE TITLE
Freeing memory

### ISSUE THREAD

**Issue by plajos1** [salylgw/gammath_spot#4]
**Describe the bug**

**To Reproduce**

**Expected behavior**

**Screenshots**

**Desktop (please complete the following information):**

**Smartphone (please complete the following information):**

**Additional context**

**Comment by plajos1** [comment]
I have tried your software many times and even made some suggestions earlier. 

I found that the program eats up all memories in my machine. I am running the program with i7 16GB W8.1 from two batch files:

python gammath_stocks_data_scraper.py watchlist0.csv > log_scraper0.txt

python gammath_stocks_analyzer_and_scorer.py watchlist0.csv > log_scorer0.txt

If Ieave the programs as is, then not even the first line could be done properly. The laptop crashes due to memory problems. To overcome this I have always modified the number of cores to used whenever you have published a new version:

cores_to_use = int(core_count/2)

Without it the program eats up all the 16GB during the scraping! I am trying to analyse 127 tickers in one run at the moment. I can see that the program is analysing only a certain number (equals to core_to_use) at a time. It means that the number of stocks should not be the problem. This means that the memory not freed up properly between the tickers. Maybe the temporary data should be saved between the tickers and concatenated in a final run? It is also strange that the reducing of cpu_to_use can solve this...

I want to compare your results to my other ideas: maybe I should split up my database and feed the gammath spot with 32 stock length of watchlists and sum up the separate results?

**Comment by salylgw** [comment]
Interesting.... instead of using "cores_to_use = int(core_count/2)" is it
possible for you to try the following:
----
import os
cores_to_use = len(os.sched_getaffinity(0))
----

I'm suspecting that the number of usable CPUs might be incorrect. Using the
above should fix the problem if that was the case.

You are right in that the number of tickers processed at a time are equal
to cores_to_use because I'm running it in parallel on as many cores as
possible. I think all resources should get freed during the join. Let's see
if the above change fixes the problem you are experiencing.

Thanks for bringing this to my attention.

**Comment by plajos1** [comment]
Tested it. Unfortunately it makes an error:

module 'os' has no attribute 'sched_getaffinity'

I have seen on the net that this is an issue on Windows based python for quite a long time.
(since 3.6)

os.cpu_count() gives me 8 on my i7, but it only has 4 cores. The 8 probably come from the number of threads. Maybe all core can be used as a core, but in a multitasking environment I have to reduce the number of actually used by the process to avoid crashes. 

I will also test if the 'sched_getaffinity' function is working properly on ubuntu...

br.

**Comment by salylgw** [comment]
Thanks. Yeah, compatibility reasons is one of the reasons why I'm not using
that API in the code. Few more comments:

1) BTW, you can use a container image (if that is possible in your
environment. I've posted those instructions in README.md) just to be able
to have the same results at least as a quick check.
2) I remember that I had seen this type of issue (count is n but it
works for (n/2)+1) in another use case. Hyperthreading (like you mentioned)
might be active. As I saw in that use case, when I was launching browser
type applications, it was seeing the problem you mentioned. At the time, I
had switched to using urllib2 for basic scraping and then it was allowing
me to use all cores. I suppose you are seeing this issue on the code "as
is". If not then please let me know.
3) I suppose the quick local workaround of setting core count to n/2 or
(n/2)+1 works for you and isn't a show stopper. Please let me know if it is
otherwise. I understand it might be at a frustrating speed.
4) At then time when I used to see this n/2 core count issue, I had found a
package on PyPI that is supposed to return usable CPU count but didn't
check further as sched_getaffinity did work for me after some Python v3.x.
5) In the mean time, you have rightly pointed out the memory usage issue.
So let me make a few comments about that as well:
a) Scraper will try to get 10Y history the first time you run it for the
symbol. Subsequent runs are only supposed to get the missing parts
(assuming you are using the latest code). So, I think in subsequent runs,
memory usage should be substantially less.
b) Pandas mostly works with in-memory data structures so if I truncate some
columns from the stock history that are not used in actual analysis then it
might show you some memory usage improvements. I can do this on the side.
I'm not sure if it will add a lot of value (simply because after
join/close, all resources should be freed before the next iteration of
parallel processes starts) but it will add some value so I've avoided using
chunks, dask pkg etc. Reducing the use of columns in the stock history
should show noticeable difference for analyzer and projector runs.
c) Historian tries to generate ~5Y worth of scores history so first time it
runs, it will show a lot of memory usage. However, the memory footprint
improvement in analyzer should have a direct impact on historian tool also.
Subsequent runs of historian is only supposed run for the missing history
so this should show far fewer iterations.
d) Backtesting can also see improvement in memory usage with changes
related to using truncated data frame.

Thanks,
Salyl

**Comment by salylgw** [comment]
I made small change to reduce the memory used by data frames during processing. Its committed. I'll include in the next release after some more testing.

Thanks,
Salyl

**Comment by salylgw** [comment]
I added "get_usable_cpu_count" (
https://github.com/salylgw/gammath_spot/blob/main/gammath_spot/gammath_utils.py)
along with few other changes that should make it easier to maintain and
configure.

BTW, please note that in my previous response, it should be "substantially
less time" instead of "substantially less memory" during subsequent runs of
scraper and historian.

Best regards,
Salyl

**Comment by plajos1** [comment]
I have tested the "get_usable_cpu_count" method the same day you answered. Unfortunately it does not work on windows: "module 'os' has no attribute 'get_usable_cpu_count' ". I tried to find a go-around since then, but there was no any solution. 

The number of cores should be correct otherwise because when I run the program with the 8 cores (reported in the original method) it works fine: in the subdirectory of "tickers" eight files are written simultaneously.  With a reduced number of tickers in the "watchlist" it is capable to finish the task. With larger watchlists the system brokes with memory message. 

I think that the history handling you have made will be a help. Otherwise the reducing of CPU cores used maybe necessary in a Windows environment to avoid the crash of the operating environment.

**Comment by salylgw** [comment]
The change I committed is slightly different in that I added a try/catch
that allows getting the CPU count that will work on your system Apparently,
this is a problem also on some MACs. Ever since I started using docker,
these differences had fallen into oblivion. I'll release it soon.

Thanks,
Salyl

**Comment by plajos1** [comment]
I am looking forward to it...

**Comment by salylgw** [comment]
v9.1 is released. When you get a chance, please let me know if the issue
persists or is resolved.

Thanks,
Salyl

**Comment by plajos1** [comment]
Yes it is working fine! Watching it during the run I have noticed that the program is using four cores. (well my i7 has four cores!) It seems that the core_count in python gave the number of threads instead of cores and this could be one of the causes resulted in memory overload on my laptop. 

I have never tested the program earlier with more than 127 tickers. To check if the modifications were effective in memory usage I have raised the number of stocks to 250! During the process the the max memory usage was about 34%, with a Chrome with six open tags running in the background. So all told it seems that you have solved both the core_count and the memory usage issues. Maybe the core/thread count caused a misery of intermittent data handling which cased the whole issue, because with my go-around solution of halving the reported core_count during the run was also capable of running without the memory issues. 

congrats...

br.: Louis

**Comment by salylgw** [comment]
Cool! Thanks for the update. If you ever want to try changing the cores
then it can be done in one single location.

Best regards,
Salyl

**Comment by salylgw** [comment]
Resolved in v9.1.

**Comment by salylgw** [comment]
Please do start a new issue for any portability or any concern.

Thanks,
Salyl
